peps/pep-0810.rst

```diff
@@ -0,0 +1,1345 @@
+
+A module may contain a ``__lazy_modules__`` attribute, which is a list of
+module names (strings) to make *potentially lazy* (as if the ``lazy`` keyword was
+used). This attribute is checked on each ``import`` statement to determine whether
+the import should be made *potentially lazy*.
+
+The normal (non-lazy) import statement will check the global lazy imports
+flag. If it is "enabled", all imports are *potentially lazy* (except for
+imports that can't be lazy, as mentioned above.)
+
+If the global lazy imports flag is set to "disabled", no *potentially lazy*
+import is ever imported lazily, and the behavior is equivalent to a regular
+import statement: the import is *eager* (as if the lazy keyword was not used).
+
+For a *potentially lazy* import, the lazy imports filter (if set) is called with
+the name of the module doing the import, the name of the module being
+imported, and (if applicable) the fromlist. If the lazy import filter returns
+``True``, the *potentially lazy* import becomes a lazy import. Otherwise, the
+import is *not* lazy, and the normal (eager) import continues.
+
+Lazy import mechanism
+---------------------
+
+When an import is lazy, ``__lazy_import__`` is called instead of
+``__import__``. ``__lazy_import__`` has the same function signature as
+``__import__``. It adds the module name to ``sys.lazy_modules``, a set of
+module names which have been lazily imported at some point (primarily for
+diagnostics and introspection), and returns a "lazy module object."
+
+The implementation of ``from ... import`` (the ``IMPORT_FROM`` bytecode
+implementation) checks if the module it's fetching from is a lazy module
+object, and if so, returns a lazy object for each name instead.
+
+The end result of this process is that lazy imports (regardless of how they
+are enabled) result in lazy objects being assigned to global variables.
+
+Lazy module objects do not appear in ``sys.modules``, they're just listed in
+the ``sys.lazy_modules`` set. Under normal operation lazy objects should
+only end up stored in global variables, and the common ways to access those
+variables (regular variable access, module attributes) will resolve lazy
+imports ("reify") and replace them when they're accessed.
+
+It is still possible to expose lazy objects through other means, like
+debuggers. This is not considered a problem.
+
+Reification
+-----------
+
+When a lazy object is first used, it needs to be reified. This means
+resolving the import at that point in the program and replacing the lazy
+object with the concrete one. Reification imports the module in the same way
+as it would have been if it had been imported eagerly, barring intervening
+changes to the import system (e.g. to ``sys.path``, ``sys.meta_path``,
+``sys.path_hooks`` or ``__import__``).
+
+Reification still calls ``__import__`` to resolve the import. Once the
+module is reified, it's removed from ``sys.lazy_modules``. When a package is
```

Comment:
Even if it's also lazily imported somewhere else? Text below suggests that reification doesn't affect other instances of the lazy module that may live elsewhere.

Reply:
Yes, it's removed when the import becomes concrete (when it's loaded and added to `sys.modules`), not when all lazy objects are reified (which may never happen). It's the set of module names that are currently *not* loaded but have been lazily imported. I've clarified the text.